CMake: allow use of system example-robot-data

nim65s opened the pull request:
first step for #2575

nim65s committed:
CMake: allow use of system example-robot-data

nim65s committed:
Nix: add example-robot-data as dependency

nim65s committed:
tests / examples: update example-robot-data use

nim65s committed:
add changelog entry

nim65s committed:
ruff check --fix .

nim65s committed:
fix RUF046

nim65s committed:
ruff: ignore RUF048

```
examples/meshcat-viewer-octree.py:13:10: RUF048 `__version__` may contain non-integral-like elements
   |
11 | from pinocchio.visualize import MeshcatVisualizer
12 |
13 | if tuple(map(int, fcl.__version__.split("."))) >= (3, 0, 0):
   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ RUF048
14 |     with_octomap = fcl.WITH_OCTOMAP
15 | else:
   |
```

nim65s committed:
CMake: fix PINOCCHIO_MODEL_DIR

nim65s committed:
CMake: allow user-specified EXAMPLE_ROBOT_DATA_MODEL_DIR

nim65s committed:
CMake: add missing links to Boost::filesystem

nim65s commented:
I don't understand why codacy is complaining

jcarpent commented:
@nim65s Why do you need to link to Boost.Filesystem?

nim65s commented:
To get mesh dir, eg.
```cpp
const std::string model_path = EXAMPLE_ROBOT_DATA_MODEL_DIR;
const std::string mesh_dir = boost::filesystem::path(model_path).parent_path().parent_path().string();
```

There are plenty of other solutions (the good one being requiring C++17 to get std::filesystem, but this can wait for pinocchio v4).
Boost::filesystem seems to be already used in other places, so I just followed current use.

I don't need the CMake part about Boost::filesystem, it was working as well without, but I tried to please codacy. And that failed.

ManifoldFR commented:
> There are plenty of other solutions (the good one being requiring C++17 to get std::filesystem, but this can wait for pinocchio v4). Boost::filesystem seems to be already used in other places, so I just followed current use.

Oh yeah, we can definitely put upgrading from Boost::filesystem to std for Pinocchio 4 since we'll go C++17. We should put that on the issue tracker for Pinocchio 4 changes for modernisation, along with stuff like boost::optional and so on.

jorisv commented on CMakeLists.txt:
```diff
@@ -161,6 +155,26 @@ option(CHECK_RUNTIME_MALLOC "Check if some memory allocations are performed at r
+    string(REPLACE "EXAMPLE_ROBOT_DATA_MODEL_DIR=" "" EXAMPLE_ROBOT_DATA_MODEL_DIR_QUOTES
+                   "${EXAMPLE_ROBOT_DATA_MODEL_DIR_DEF}")
+    string(REPLACE "\"" "" EXAMPLE_ROBOT_DATA_MODEL_DIR "${EXAMPLE_ROBOT_DATA_MODEL_DIR_QUOTES}")
+  endif()
```
Add a message to show which directory we will use

nim65s replied:
fixed in 370a3320b7ea6865f35330e587a2cc03c0454680

jorisv commented on examples/CMakeLists.txt:
```diff
@@ -37,6 +37,13 @@ function(ADD_PINOCCHIO_CPP_EXAMPLE EXAMPLE)
     PRIVATE ${EXAMPLE_PRIVATE_DEFINITIONS} PINOCCHIO_MODEL_DIR="${PINOCCHIO_MODEL_DIR}"
             COAL_DISABLE_HPP_FCL_WARNINGS)
```
```suggestion
    PRIVATE ${EXAMPLE_PRIVATE_DEFINITIONS} COAL_DISABLE_HPP_FCL_WARNINGS)
```

nim65s replied:
fixed in a072522497c2c96f1d3cc62f885b1fc00e48a238

jorisv commented on unittest/CMakeLists.txt:
Remove `PINOCCHIO_MODEL_DIR=\"${PINOCCHIO_MODEL_DIR}\"` ligne 61.

nim65s replied:
This is still used in multiple tests, for `simple_humanoid` urdf & sdf

jorisv replied:
And the point

jorisv commented on unittest/packaging/cmake/CMakeLists.txt:
```diff
@@ -1,10 +1,13 @@
 
 set(PINOCCHIO_MODEL_DIR "${PROJECT_SOURCE_DIR}/../../../models")
 add_definitions('-DPINOCCHIO_MODEL_DIR=\"${PINOCCHIO_MODEL_DIR}\"')
+set(EXAMPLE_ROBOT_DATA_MODEL_DIR "\"${PINOCCHIO_MODEL_DIR}/example-robot-data/robots\"")
+add_definitions('-DEXAMPLE_ROBOT_DATA_MODEL_DIR=\"${EXAMPLE_ROBOT_DATA_MODEL_DIR}\"')
```
Why doing that if you don't update .cpp file ? (they still use PINOCCHIO_MODEL_DIR)

nim65s replied:
fixed in 4793f45ed5aeffcd875b88c8940979c431ad46bc

jorisv reviewed (changes requested):
You also have to modify benchmarks.

nim65s committed:
benchmarks: update example-robot-data use

nim65s committed:
examples: update example-robot-data use

nim65s committed:
CMake: log system example-robot-data path

jorisv commented on CMakeLists.txt:
```diff
@@ -161,6 +155,27 @@ option(CHECK_RUNTIME_MALLOC "Check if some memory allocations are performed at r
+   OR BUILD_TESTING)
+  # example-robot-data is required
+  if(EXAMPLE_ROBOT_DATA_MODEL_DIR)
+    message(INFO "Using provided example-robot-data at: ${EXAMPLE_ROBOT_DATA_MODEL_DIR}")
```
```suggestion
    message(STATUS "Using provided example-robot-data at: ${EXAMPLE_ROBOT_DATA_MODEL_DIR}")
```

jorisv replied:
Also, EXAMPLE_ROBOT_DATA_MODEL_DIR provided by user should be an absolute path.

We can use `file(REAL_PATH <path> <out-var>EXPAND_TILDE)` to ensure that.

jorisv commented on CMakeLists.txt:
```diff
@@ -161,6 +155,27 @@ option(CHECK_RUNTIME_MALLOC "Check if some memory allocations are performed at r
+    string(REPLACE "EXAMPLE_ROBOT_DATA_MODEL_DIR=" "" EXAMPLE_ROBOT_DATA_MODEL_DIR_QUOTES
+                   "${EXAMPLE_ROBOT_DATA_MODEL_DIR_DEF}")
+    string(REPLACE "\"" "" EXAMPLE_ROBOT_DATA_MODEL_DIR "${EXAMPLE_ROBOT_DATA_MODEL_DIR_QUOTES}")
+    message(INFO "Using system example-robot-data at: ${EXAMPLE_ROBOT_DATA_MODEL_DIR}")
```
```suggestion
    message(STATUS "Using system example-robot-data at: ${EXAMPLE_ROBOT_DATA_MODEL_DIR}")
```

jorisv committed:
unittest: Don't mix local PINOCCHIO_MODEL_DIR and EXAMPLE_ROBOT_DATA_MODEL_DIR

jorisv committed:
benchmark: Fix boost/filesystem inculde

nim65s committed:
CMake: improve EXAMPLE_ROBOT_DATA_MODEL_DIR ux

Works with `-DEXAMPLE_ROBOT_DATA_MODEL_DIR=~/here` and
`-DEXAMPLE_ROBOT_DATA_MODEL_DIR=../here` (big warning in the second
case, but that is fine)

Co-authored-by: Joris Vaillant <778707+jorisv@users.noreply.github.com>

ManifoldFR commented:
Great work people!!
